File: backend/app/api/v1/endpoints/assets.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query

from app.schemas.asset import AssetCreate, AssetResponse, AssetUpdate
from app.schemas.common import MessageResponse, PaginatedResponse, PaginationParams
from app.services.asset_service import AssetService

router = APIRouter(prefix="/assets", tags=["assets"])


@router.post("", response_model=AssetResponse, status_code=201)
async def create_asset(schema: AssetCreate, svc: AssetService = Depends()):
    return await svc.create(schema)


@router.get("", response_model=PaginatedResponse[AssetResponse])
async def list_assets(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    search: str | None = Query(None),
    status: str | None = Query(None),
    type: str | None = Query(None),
    sort_by: str | None = Query(None),
    sort_order: str = Query("desc", pattern="^(asc|desc)$"),
    svc: AssetService = Depends(),
):
    params = PaginationParams(page=page, page_size=page_size, sort_by=sort_by, sort_order=sort_order)
    filters = {k: v for k, v in {"status": status, "type": type}.items() if v}
    return await svc.list(params, search=search, filters=filters)


@router.get("/tree", response_model=list[AssetResponse])
async def get_asset_tree(svc: AssetService = Depends()):
    return await svc.get_tree()


@router.get("/{asset_id}", response_model=AssetResponse)
async def get_asset(asset_id: UUID, svc: AssetService = Depends()):
    result = await svc.get(asset_id)
    if not result:
        raise HTTPException(404, "Asset not found")
    return result


@router.patch("/{asset_id}", response_model=AssetResponse)
async def update_asset(asset_id: UUID, schema: AssetUpdate, svc: AssetService = Depends()):
    result = await svc.update(asset_id, schema)
    if not result:
        raise HTTPException(404, "Asset not found")
    return result


@router.delete("/{asset_id}", response_model=MessageResponse)
async def delete_asset(asset_id: UUID, svc: AssetService = Depends()):
    deleted = await svc.delete(asset_id)
    if not deleted:
        raise HTTPException(404, "Asset not found")
    return MessageResponse(message="Asset deleted")

File: frontend/src/features/assets/api/assets.ts
import { api } from "@/lib/api";
import type { Asset, AssetCreate, AssetUpdate, PaginatedResponse } from "@/types";

export interface AssetFilters {
  status?: string;
  type?: string;
  search?: string;
  page?: number;
  page_size?: number;
  sort_by?: string;
  sort_order?: "asc" | "desc";
}

export const assetsApi = {
  list: (filters?: AssetFilters) => {
    const params: Record<string, string> = {};
    if (filters?.status) params.status = filters.status;
    if (filters?.type) params.type = filters.type;
    if (filters?.search) params.search = filters.search;
    if (filters?.page) params.page = String(filters.page);
    if (filters?.page_size) params.page_size = String(filters.page_size);
    if (filters?.sort_by) params.sort_by = filters.sort_by;
    if (filters?.sort_order) params.sort_order = filters.sort_order;
    return api.get<PaginatedResponse<Asset>>("/api/v1/assets", params);
  },

  get: (id: string) => api.get<Asset>(`/api/v1/assets/${id}`),

  create: (data: AssetCreate) => api.post<Asset>("/api/v1/assets", data),

  update: (id: string, data: AssetUpdate) => api.patch<Asset>(`/api/v1/assets/${id}`, data),

  delete: (id: string) => api.delete<void>(`/api/v1/assets/${id}`),
};
